test/PerformanceTests/SerializationComparisonsTests/Lib/ODataMessageWriterAsyncPayloadWriter.cs

```diff
@@ -92,14 +94,30 @@ public async Task WritePayloadAsync(IEnumerable<Customer> payload, Stream stream
                 // start write homeAddress
                 await writer.WriteStartAsync(homeAddressInfo);
 
-                var homeAddressResource = new ODataResource
+                ODataResource homeAddressResource;
+                if (includeRawValues)

```

Comment:
I don't know if there was a specific reason why in _test/PerformanceTests/SerializationComparisonsTests/Lib/DataModel.cs_ you added `Misc` property in between `City` and `Street` but if you were to add it to the bottom, I believe you'd be able to avoid duplicate code in this file **plus other files with similar duplicated code**:
```csharp
var homeAddressPropertes = new List<ODataProperty>
{
    new ODataProperty { Name = "City", Value = customer.HomeAddress.City },
    new ODataProperty { Name = "Street", Value = customer.HomeAddress.Street }
};

if (includeRawValues)
{
    homeAddressProperties.Add(new ODataProperty { Name = "Misc", Value = new ODataUntypedValue() { RawValue = $"\"{customer.HomeAddress.Misc}\"" } })
}

var homeAddressResource = new ODataResource
{
    Properties = homeAddressProperties
}
```
Even with `Misc` sandwiched between `City` and `Street`, you should be able to do this:
```csharp
var homeAddressPropertes = new List<ODataProperty>
{
    new ODataProperty { Name = "City", Value = customer.HomeAddress.City }
};

if (includeRawValues)
{
    homeAddressProperties.Add(new ODataProperty { Name = "Misc", Value = new ODataUntypedValue() { RawValue = $"\"{customer.HomeAddress.Misc}\"" } })
}

homeAddressProperties.Add(
    new ODataProperty { Name = "Street", Value = customer.HomeAddress.Street });

var homeAddressResource = new ODataResource
{
    Properties = homeAddressProperties
}
```

This comment applies to both sync and async sections of similar code in the following files:
- _test/PerformanceTests/SerializationComparisonsTests/Lib/ODataMessageWriterPayloadWriter.cs_
- _test/PerformanceTests/SerializationComparisonsTests/Lib/ODataMessageWriterAsyncPayloadWriter.cs_

Reply:
Thanks. There's no reason to have it in between. Add the property first, then added the conditional statement later in order to be able to run tests without raw values.

Reply:
By the way, the reason I had the if statement outside of the loop was to avoid calling the if statement in each iteration since the value of `includeRawValues` doesn't change inside the loop. I thought having the if statement inside the loop means the benchmark test would pay the penalty of the if statement even when the test run doesn't include raw values. Maybe this is something that the system can optimize on its own via branch prediction. I didn't compare the benchmarks between having the if statement in the loop and outside to validate whether it makes a difference.

Reply:
I have modified the `foreach (address in customer.Addresses)` code to move the if statement inside the loop and noticed the performance slow down by 10s on the `ODataMessageWriter` sync and async tests. However, this could be a fluke. But I've just realized your question was based on `customer.HomeAddress` and not `customer.Addresses`.

Reply:
I think the perf degradation was noise, I've ran it again and the results were closer to the original run.

Reply:
That said, there's still probably a minor cost, initially I was using arrays to store the properties, now I'd be using `List` and `List.Add()`

Reply:
There's an increase in memory allocations by about 3MB and this does not appear to be a fluke (it's consistent across multiple runs). But it seems too big a difference to be explained by changing from arrays to lists.

Reply:
I think I should keep the initial code at the moment since the memory allocations has reduced when I reverted it to what it was.

Reply:
@habbes Curious, why does `includeRawValues` have to be an option? Why can't you _just_ update the existing benchmark to include raw values?

Reply:
The reason I made it an option is that it made the benchmarks slower (partly because we write more data as result of additional property and partly because we have to manually write a semicolon keep track of related flags), which makes it harder to compare to older benchmarks that did not include raw values. I also wanted to ensure that changes made to address raw values did not affect performance when not using raw values. My assumption is that raw values are an edge case and so people who do not use raw values should not incur a noticeable performance penalty because we added code that handles raw values. So, I wanted to be able to see the perf with and without raw values.

Reply:
Another reason the raw value benchmarks are slower is because I'm doing the string interpolation inside the benchmark and not when the benchmark data is generated.

Reply:
https://github.com/OData/odata.net/pull/2527#discussion_r1035540039
"_people who do not use raw values should not incur a noticeable performance penalty because we added code that handles raw values_" - By this statement I assume you're only referring to performance penalty in respect to running the benchmarks, not penalty in using the actual library..?

Reply:
Performance penalty with respect to the benchmarks. But for me to know that performance for scenarios without raw values was not affected after the change, I needed a way to run benchmarks without writing raw values.